docs/migration-2.md

```diff
@@ -0,0 +1,119 @@
+---
+comments: true
+---
+
+# Migrating from `g2p` 1.0
+
+The `g2p` 2.0 release introduces a number of improvements and changes
+which, unfortunately, are incompatible with mappings and Python code
+written for the previous version.  We have tried to describe them here
+with the changes you will need to make to your code and data.
+
+## Mapping configurations have changed (for the better)
+
+The configurations for mappings (which you'll find in
+`g2p/mappings/langs/*/config-g2p.yaml`) are now validated with a [YAML
+Schema](https://raw.githubusercontent.com/roedoejet/g2p/main/g2p/mappings/.schema/g2p-config-schema-2.0.json).
+If you use an editor like [Visual Studio
+Code](https://code.visualstudio.com/), the names of fields will be
+autocompleted and some warnings will be shown for possible values.
+This also works with [GNU Emacs](https://www.gnu.org/software/emacs/)
+using [Eglot](https://joaotavora.github.io/eglot/) or
+[lsp-mode](https://emacs-lsp.github.io/lsp-mode/) and any other editor
+that supports the [Language Server
+Protocol](https://microsoft.github.io/language-server-protocol/)
+and/or [SchemaStore](https://www.schemastore.org/json/).  Some
+varieties of VIM are known to work, for instance.
+
+In order for this magic to work, we needed to give the configuration
+files a somewhat more meaningful name than `config.yaml`, so they must
+now be called `config-g2p.yaml`.  In addition some fields have changed
+names to reflect the fact that they refer to *files* and not the
+actual rules themselves:
+
+- `mapping` is now `rules_path`
+- `abbreviations` is now `abbreviations_path`
+
+The mappings themselves should be compatible with the previous
+version, please let us know if you encounter any problems.
+
+## Submodules of `g2p` must be imported explicitly
+
+Previously, when you called `import g2p`, it imported absolutely
+everything, which caused the command-line interface (and probably your
+program too) to start up very, very slowly.
+
+If you simply use the public and documented `make_g2p` API, this will
+not change anything, but if you relied on internal classes and
+functions from `g2p.mappings`, `g2p.transducer`, etc, then you can no
+longer depend on them being also accessible in the top-level `g2p`
+package.  For example, you will need to make this sort of change:
+
+```diff
+- from g2p import Mapping, Transducer, make_tokenizer
++ from g2p.mappings import Mapping
++ from g2p.transducer import Transducer
++ from g2p.mappings.tokenizer import make_tokenizer
```

Comment:
`make_tokenizer` is also part of the official documented API, and it should be imported as `from g2p import make_tokenizer`. You'll notice in `g2p/__init__.py` that it's just a dummy wrapper, but it has the major advantage of deferring the import to actual use, which means this is the recommended way to import that function for packages using g2p as an external module.

This section is correct for other things, like `Mapping` and `Transducer`.